### PR TITLE
Feat/authz provider

### PR DESCRIPTION
**Pull request by fantix**
### New Features
- Added authz provider support

### Deployment changes
- One-off: Update authz_provider to the correct values

### CODE REVIEW

File: arborist/server.go [hunk 1]
```diff
@@ -1260,12 +1260,14 @@ func (server *Server) handleGroupCreate(w http.ResponseWriter, r *http.Request,
...
 		return
 	}
-	var errResponse *ErrorResponse
-	if r.Method == "PUT" {
-		errResponse = transactify(server.db, group.overwriteInDb)
-	} else {
-		errResponse = transactify(server.db, group.createInDb)
-	}
+	authzProvider := getAuthZProvider(r)
+	errResponse := transactify(server.db, func(tx *sqlx.Tx) *ErrorResponse{
```
Comment (vpsx) [inline arborist/server.go]: looooooool
I started this https://play.golang.org/p/kuLvJBd8lNL , don't know how to get past the type error but maybe you do? If you like the general idea, that is

File: arborist/group.go [hunk 2]
```diff
@@ -216,20 +214,58 @@ func (group *Group) deleteInDb(tx *sqlx.Tx) *ErrorResponse {
...
-func grantGroupPolicy(db *sqlx.DB, groupName string, policyName string) *ErrorResponse {
+func grantGroupPolicy(db *sqlx.DB, groupName string, policyName string, authzProvider sql.NullString) *ErrorResponse {
 	stmt := `
-		INSERT INTO grp_policy(grp_id, policy_id)
-		VALUES ((SELECT id FROM grp WHERE name = $1), (SELECT id FROM policy WHERE name = $2))
+		INSERT INTO grp_policy(grp_id, policy_id, authz_provider)
+		VALUES ((SELECT id FROM grp WHERE name = $1), (SELECT id FROM policy WHERE name = $2), $3)
 	`
-	_, err := db.Exec(stmt, groupName, policyName)
+	_, err := db.Exec(stmt, groupName, policyName, authzProvider)
```
Comment (vpsx) [inline arborist/group.go]: same question as for grantUserPolicy

File: arborist/user.go [hunk 3]
```diff
@@ -186,13 +187,13 @@ func (user *User) deleteInDb(db *sqlx.DB) *ErrorResponse {
 	return nil
 }
 
-func grantUserPolicy(db *sqlx.DB, username string, policyName string, expiresAt *time.Time) *ErrorResponse {
+func grantUserPolicy(db *sqlx.DB, username string, policyName string, expiresAt *time.Time, authzProvider sql.NullString) *ErrorResponse {
 	stmt := `
-		INSERT INTO usr_policy(usr_id, policy_id, expires_at)
-		VALUES ((SELECT id FROM usr WHERE name = $1), (SELECT id FROM policy WHERE name = $2), $3)
+		INSERT INTO usr_policy(usr_id, policy_id, expires_at, authz_provider)
```
[the remark this answers is not in the document]
Comment (vpsx) [reply arborist/user.go]: Yes, desired. Not assuming authz_provider always present. For compatibility with old clients. But we are going to manually update authz_provider field in db after migration. If authz_provider not valid (string is null), then sql stmt just inserts NULL.

File: arborist/group.go [hunk 4]
```diff
@@ -133,12 +134,7 @@ func (group *Group) policies(tx *sqlx.Tx) ([]PolicyFromQuery, error) {
 	return policies, nil
 }
 
-func (group *Group) createInDb(tx *sqlx.Tx) *ErrorResponse {
-	// First, insert permissions if they don't exist yet. If they don't exist
```
[the remark this answers is not in the document]
Comment (vpsx) [reply arborist/group.go]: Yes.

File: arborist/server.go (hunk 1 above)
Comment (vpsx) [reply arborist/server.go]: go T^T

File: arborist/group.go [hunk 5]
```diff
@@ -216,20 +214,58 @@ func (group *Group) deleteInDb(tx *sqlx.Tx) *ErrorResponse {
...
+			if authzProvider.Valid {
+				msg = fmt.Sprintf("failed to clear %s grp_policy for %s", authzProvider.String, group.Name)
+			} else {
+				msg = fmt.Sprintf("failed to clear grp_policy for %s", group.Name)
+			}
+			return newErrorResponse(msg, 500, &err)
+		}
+
+		return group.attachUsrAndPolicy(tx, groupID, authzProvider)
+	} else {
```
[the remark this answers is not in the document]
Comment (fantix) [reply arborist/group.go]: done

File: arborist/server_test.go [hunk 6]
```diff
@@ -1512,36 +1513,49 @@ func TestServer(t *testing.T) {
...
+			test := func(authzProvider string, expected bool, msg string) {
+				w := httptest.NewRecorder()
+				url := fmt.Sprintf("/user/%s/policy/%s", username, policyName)
+				req := newRequest("DELETE", url, nil)
+				req.Header.Add("X-AuthZ-Provider", authzProvider)
+				handler.ServeHTTP(w, req)
+				if w.Code != http.StatusNoContent {
+					httpError(t, w, "couldn't revoke policy")
+				}
+				// look up user again and check that policy is still there
```
[the remark this answers is not in the document]
Comment (fantix) [reply arborist/server_test.go]: yes!

**Pull request merged**